linkerd/router/src/cache.rs

```diff
@@ -0,0 +1,89 @@
+use crate::NewService;
+use ahash::AHashMap;
+use parking_lot::Mutex;
+use std::{fmt::Debug, hash::Hash, marker::PhantomData, sync::Arc};
+
+/// A [`NewService`] that lazy builds stacks for each `K`-typed key.
+#[derive(Debug)]
+pub struct NewCache<K, N> {
+    inner: N,
+    _marker: PhantomData<fn(K)>,
+}
+
+/// A [`NewService`] that lazily builds an inner `S`-typed service for each
+/// `K`-typed target.
```

Comment:
should maybe explicitly state that the `S`-typed service is only built once per `K` and then cloned?

Reply:
maybe also worth documenting that this cache never evicts?